### ISSUE TITLE
typeid equality in CheckedDowncastToActualType() failing on OS X (Shared Library + Release) - Ceres Solver

### ISSUE THREAD

**Issue by alexsmac** [google/googletest#595]
In the Ceres Solver, we recently updated our gtest to master.  Since doing so on OS X (Xcode 6 & 7), only when Ceres is built as a shared library (via BUILD_SHARED_LIBS=ON) with CMAKE_BUILD_TYPE=Release (the default), we are seeing two tests failing (gradient_checking_cost_function_test & levenberg_marquardt_strategy_test) in: CheckedDowncastToActualType().  Specifically this line:

GTEST_CHECK_(typeid(*base) == typeid(Derived));

even though the names via: typeid(X).name() _are_ equal, i.e. the following test passes:

GTEST_CHECK_(std::string(typeid(*base).name()) == std::string(typeid(Derived).name()));

This does not happen when Ceres (and its internal gtest library) are compiled statically (irrespective of the CMAKE_BUILD_TYPE) or if compiled in Debug even if they are compiled as shared libraries.  All tests pass on 14.04 with both Clang and GCC.

This appears to be a manifestation of the problem detailed here: https://whatofhow.wordpress.com/2015/03/17/odr-rtti-dso/ although it appears our update of gtest (rather than a change from Xcode 6 to 7) is the issue, as reverting Ceres to the previous commit fixes everything.

In light of the issues with typeid checks with shared libraries, is there any objection to updating gtest to compare the names as above instead of relying on typeid::operator== ?  I think the performance penalty would be small, although it seems the safest way to perform the comparison whilst accounting for shared libraries.

Thanks!

-Alex

**Comment by BillyDonahue** [comment]
I have a few questions.

You said Ceres has an INTERNAL gtest library?
Can I ask why that's necessary? What does "internal" mean?

typeid().name() isn't guaranteed to have any particular value, and it's allowed to be empty.
Are you sure that's not why your == test is passing?

**Comment by sandwichmaker** [comment]
we just have a copy of googletest + googlemock generated using the fusion scripts inside ceres.

**Comment by sandwichmaker** [comment]
internal just means that we do not expose it as part of our API, we just use it for our own tests.

**Comment by alexsmac** [comment]
Hi Billy - typeid().name() here has reasonable values here at least, e.g: N7testing8internal11ThreadLocalIPNS_8SequenceEE11ValueHolderE

**Comment by BillyDonahue** [comment]
Thanks for the answers.

I'm still upset by the suggestion to change gtest to rely on names though, because typeid().name() isn't required to have reasonable values or any value at all.

I'm trying to study the article.
RTLD_LOCAL or RTLD_GLOBAL?

Can you rely on dynamic_cast instead of typeid?

**Comment by alexsmac** [comment]
After some more digging, I found the following:

The trigger for the failure when building in Release, is that (only in Release) do we enable link-time optimisation (if using Clang, and if the linker supports it).  If I remove -flto, then all tests pass without issue.

I've tried the nm tricks described here: http://www.russellmcc.com/posts/2013-08-03-rtti.html to look for duplicate typeinfo symbols, but I can't find any (with or without LTO).  I also checked (as suggested by the article) that we aren't passing any -fvisibility flags, and we aren't for OS X (or Linux / Windows), and we aren't - although we do for iOS which I think they are required.

I've double checked via LLDB that type_info::operator== is using the address of the object rather than comparing the name on OS X as seen here:

https://github.com/llvm-mirror/libcxx/blob/60784f62deefe20952cabe220eb3f14b20c5e659/include/typeinfo#L127

and indeed it is, i.e. _LIBCPP_NONUNIQUE_RTTI_BIT is not defined.  However, as noted here: http://lists.llvm.org/pipermail/cfe-commits/Week-of-Mon-20140324/102234.html this is not true for iOS ARM64 builds, where  operator== actually uses a string comparison of the name from type_info.  GCC has similar conditions for other cases too: https://gcc.gnu.org/onlinedocs/libstdc++/latest-doxygen/a01567_source.html#l00088

Billy, I also checked that dynamic_cast<Derived*>(base) returns a non-NULL pointer, and it does, even when operator== fails.  So at least that is still working.

At this point, I don't think this is an issue with gtest - this _looks_ like an LTO fail.  I concur with Billy that gtest should be using type_info::operator==(), and relying on that to select whether to perform an address or string comparison as required.
